diff --git a/clang/BUILD.bazel b/clang/BUILD.bazel
index a17b0b54a49b..12aa69678e2c 100644
--- a/clang/BUILD.bazel
+++ b/clang/BUILD.bazel
@@ -1900,6 +1900,11 @@ genrule(
     toolchains = [":workspace_root"],
 )
 
+filegroup(
+    name = "builtin_headers_files",
+    srcs = builtin_headers,
+)
+
 cc_library(
     name = "frontend",
     srcs = glob([
diff --git a/llvm/config.bzl b/llvm/config.bzl
index 2309175d04a9..4b3c6b2791db 100644
--- a/llvm/config.bzl
+++ b/llvm/config.bzl
@@ -28,8 +28,8 @@ def native_arch_defines(arch, triple):

 posix_defines = [
     "LLVM_ON_UNIX=1",
-    "HAVE_BACKTRACE=1",
-    "BACKTRACE_HEADER=<execinfo.h>",
+    # "HAVE_BACKTRACE=1",
+    # "BACKTRACE_HEADER=<execinfo.h>",
     r'LTDL_SHLIB_EXT=\".so\"',
     r'LLVM_PLUGIN_EXT=\".so\"',
     "LLVM_ENABLE_LLVM_EXPORT_ANNOTATIONS=1",
@@ -51,7 +51,7 @@ posix_defines = [
 linux_defines = posix_defines + [
     "_GNU_SOURCE",
     "HAVE_GETAUXVAL=1",
-    "HAVE_MALLINFO=1",
+    # "HAVE_MALLINFO=1",
     "HAVE_SBRK=1",
     "HAVE_STRUCT_STAT_ST_MTIM_TV_NSEC=1",
 ]
diff --git a/mlir/tblgen.bzl b/mlir/tblgen.bzl
index e45ba1fe0ef7..5af5db59c608 100644
--- a/mlir/tblgen.bzl
+++ b/mlir/tblgen.bzl
@@ -172,7 +172,7 @@ def _gentbl_rule_impl(ctx):
         # Make sure action_env settings are honored so the env is the same as
         # when the tool was built. Important for locating shared libraries with
         # a custom LD_LIBRARY_PATH.
-        use_default_shell_env = True,
+        # use_default_shell_env = True,
         mnemonic = "TdGenerate",
     )
 
