diff --git a/llvm/tools/llvm-readtapi/CMakeLists.txt b/llvm/tools/llvm-readtapi/CMakeLists.txt
index 855a1473888f..6d83c3263c5e 100644
--- a/llvm/tools/llvm-readtapi/CMakeLists.txt
+++ b/llvm/tools/llvm-readtapi/CMakeLists.txt
@@ -17,6 +17,7 @@ add_llvm_tool(llvm-readtapi
 
   DEPENDS
   ReadTAPIOptsTableGen
+  GENERATE_DRIVER
   )
 
 if(LLVM_INSTALL_BINUTILS_SYMLINKS)
diff --git a/llvm/tools/llvm-readtapi/llvm-readtapi.cpp b/llvm/tools/llvm-readtapi/llvm-readtapi.cpp
index 7a8cad56f6f7..312f298a8625 100644
--- a/llvm/tools/llvm-readtapi/llvm-readtapi.cpp
+++ b/llvm/tools/llvm-readtapi/llvm-readtapi.cpp
@@ -17,7 +17,7 @@
 #include "llvm/Support/CommandLine.h"
 #include "llvm/Support/Error.h"
 #include "llvm/Support/FileSystem.h"
-#include "llvm/Support/InitLLVM.h"
+#include "llvm/Support/LLVMDriver.h"
 #include "llvm/Support/MemoryBuffer.h"
 #include "llvm/Support/Path.h"
 #include "llvm/Support/raw_ostream.h"
@@ -487,8 +487,7 @@ static void setStubOptions(opt::InputArgList &Args, StubOptions &Opt) {
   Opt.TraceLibs = Args.hasArg(OPT_t);
 }
 
-int main(int Argc, char **Argv) {
-  InitLLVM X(Argc, Argv);
+int llvm_readtapi_main(int Argc, char **Argv, const llvm::ToolContext &) {
   BumpPtrAllocator A;
   StringSaver Saver(A);
   TAPIOptTable Tbl;
diff --git a/llvm/BUILD.bazel b/llvm/BUILD.bazel
index c4955ed888cc..f58f7adb8b94 100644
--- a/llvm/BUILD.bazel
+++ b/llvm/BUILD.bazel
@@ -5592,15 +5592,13 @@ gentbl_cc_library(
     deps = [":OptParserTdFiles"],
 )
 
-cc_binary(
-    name = "llvm-readtapi",
-    testonly = True,
+cc_library(
+    name = "llvm-readtapi-lib",
     srcs = glob([
         "tools/llvm-readtapi/*.cpp",
         "tools/llvm-readtapi/*.h",
     ]),
     copts = llvm_copts,
-    stamp = 0,
     deps = [
         ":BinaryFormat",
         ":Object",
@@ -5612,6 +5610,13 @@ cc_binary(
     ],
 )
 
+llvm_driver_cc_binary(
+    name = "llvm-readtapi",
+    testonly = True,
+    stamp = 0,
+    deps = [":llvm-readtapi-lib"],
+)
+
 gentbl_cc_library(
     name = "TLICheckerOptsTableGen",
     strip_include_prefix = "tools/llvm-tli-checker",
diff --git a/llvm/driver.bzl b/llvm/driver.bzl
index 161ad2b1a936..43e563ee9d74 100644
--- a/llvm/driver.bzl
+++ b/llvm/driver.bzl
@@ -32,3 +32,4 @@ _TOOLS = {
     "llvm-readobj": "//llvm:llvm-readobj-lib",
+    "llvm-readtapi": "//llvm:llvm-readtapi-lib",
     "llvm-size": "//llvm:llvm-size-lib",
     "llvm-symbolizer": "//llvm:llvm-symbolizer-lib",
