diff --git a/swift/internal/swift_autoconfiguration.bzl b/swift/internal/swift_autoconfiguration.bzl
--- a/swift/internal/swift_autoconfiguration.bzl
+++ b/swift/internal/swift_autoconfiguration.bzl
@@ -40,6 +40,8 @@ load(
 )
 load(":toolchain_utils.bzl", "SWIFT_TOOLCHAIN_TYPE")
 
+SWIFT_AUTOCONFIG_DISABLE_ENV_VAR = "BAZEL_DO_NOT_DETECT_SWIFT_TOOLCHAIN"
+
 def _scratch_file(repository_ctx, temp_dir, name, content = ""):
     """Creates and returns a scratch file with the given name and content.
 
@@ -380,6 +382,15 @@ def _create_windows_toolchain(*, repository_ctx):
     )
 
 def _swift_autoconfiguration_impl(repository_ctx):
+    if repository_ctx.os.environ.get(SWIFT_AUTOCONFIG_DISABLE_ENV_VAR) == "1":
+        repository_ctx.file(
+            "BUILD",
+            "# Swift toolchain autoconfiguration was disabled by {} env variable.".format(
+                SWIFT_AUTOCONFIG_DISABLE_ENV_VAR,
+            ),
+        )
+        return
+
     repository_ctx.file(
         "BUILD",
         "\n".join([
@@ -406,7 +417,7 @@ def _swift_autoconfiguration_impl(repository_ctx):
     )
 
 swift_autoconfiguration = repository_rule(
-    environ = ["CC", "PATH", "ProgramData", "Path"],
+    environ = ["CC", "PATH", "ProgramData", "Path", SWIFT_AUTOCONFIG_DISABLE_ENV_VAR],
     implementation = _swift_autoconfiguration_impl,
     local = True,
 )
