diff --git a/libstdc++-v3/src/c++17/fs_dir.cc b/libstdc++-v3/src/c++17/fs_dir.cc
--- a/libstdc++-v3/src/c++17/fs_dir.cc
+++ b/libstdc++-v3/src/c++17/fs_dir.cc
@@ -147,7 +147,7 @@
 }
 
 const fs::directory_entry&
-fs::directory_iterator::operator*() const
+fs::directory_iterator::operator*() const noexcept
 {
   if (!_M_dir)
     _GLIBCXX_THROW_OR_ABORT(filesystem_error(
diff --git a/libstdc++-v3/src/c++17/fs_path.cc b/libstdc++-v3/src/c++17/fs_path.cc
--- a/libstdc++-v3/src/c++17/fs_path.cc
+++ b/libstdc++-v3/src/c++17/fs_path.cc
@@ -328,7 +328,7 @@
 }
 
 inline auto
-path::_List::begin() noexcept
+path::_List::begin()
 -> iterator
 {
   __glibcxx_assert(!empty());
@@ -338,7 +338,7 @@
 }
 
 inline auto
-path::_List::end() noexcept
+path::_List::end()
 -> iterator
 {
   __glibcxx_assert(!empty());
@@ -348,7 +348,7 @@
 }
 
 auto
-path::_List::begin() const noexcept
+path::_List::begin() const
 -> const_iterator
 {
   __glibcxx_assert(!empty());
@@ -358,7 +358,7 @@
 }
 
 auto
-path::_List::end() const noexcept
+path::_List::end() const
 -> const_iterator
 {
   __glibcxx_assert(!empty());
