commit e0b69f7d1737a96728b5918e121f6a4057c45437
Author: David Zbarsky <dzbarsky@gmail.com>
Date:   Mon Dec 8 13:55:15 2025 -0500

    No more genrules

diff --git a/llvm/lib/Support/BLAKE3/blake3_avx2_x86-64_unix.S b/llvm/lib/Support/BLAKE3/blake3_avx2_x86-64_unix.S
index e98893c7ef8b..d6fa51638da3 100644
--- a/llvm/lib/Support/BLAKE3/blake3_avx2_x86-64_unix.S
+++ b/llvm/lib/Support/BLAKE3/blake3_avx2_x86-64_unix.S
@@ -16,15 +16,17 @@
 #define _CET_ENDBR
 #endif
 
-#ifdef __APPLE__
-#define HIDDEN .private_extern
+#if defined(__APPLE__)
+#define HIDDEN(symbol) .private_extern symbol
+#elif defined(__ELF__)
+#define HIDDEN(symbol) .hidden symbol
 #else
-#define HIDDEN .hidden
+#define HIDDEN(symbol)
 #endif
 
 .intel_syntax noprefix
-HIDDEN _blake3_hash_many_avx2
-HIDDEN blake3_hash_many_avx2
+HIDDEN(_blake3_hash_many_avx2)
+HIDDEN(blake3_hash_many_avx2)
 .global _blake3_hash_many_avx2
 .global blake3_hash_many_avx2
 #ifdef __APPLE__
diff --git a/llvm/lib/Support/BLAKE3/blake3_avx512_x86-64_unix.S b/llvm/lib/Support/BLAKE3/blake3_avx512_x86-64_unix.S
index b4b14946de10..cf2c88329af5 100644
--- a/llvm/lib/Support/BLAKE3/blake3_avx512_x86-64_unix.S
+++ b/llvm/lib/Support/BLAKE3/blake3_avx512_x86-64_unix.S
@@ -16,21 +16,23 @@
 #define _CET_ENDBR
 #endif
 
-#ifdef __APPLE__
-#define HIDDEN .private_extern
+#if defined(__APPLE__)
+#define HIDDEN(symbol) .private_extern symbol
+#elif defined(__ELF__)
+#define HIDDEN(symbol) .hidden symbol
 #else
-#define HIDDEN .hidden
+#define HIDDEN(symbol)
 #endif
 
 .intel_syntax noprefix
-HIDDEN _blake3_hash_many_avx512
-HIDDEN blake3_hash_many_avx512
-HIDDEN blake3_compress_in_place_avx512
-HIDDEN _blake3_compress_in_place_avx512
-HIDDEN blake3_compress_xof_avx512
-HIDDEN _blake3_compress_xof_avx512
-HIDDEN blake3_xof_many_avx512
-HIDDEN _blake3_xof_many_avx512
+HIDDEN(_blake3_hash_many_avx512)
+HIDDEN(blake3_hash_many_avx512)
+HIDDEN(blake3_compress_in_place_avx512)
+HIDDEN(_blake3_compress_in_place_avx512)
+HIDDEN(blake3_compress_xof_avx512)
+HIDDEN(_blake3_compress_xof_avx512)
+HIDDEN(blake3_xof_many_avx512)
+HIDDEN(_blake3_xof_many_avx512)
 .global _blake3_hash_many_avx512
 .global blake3_hash_many_avx512
 .global blake3_compress_in_place_avx512
diff --git a/llvm/lib/Support/BLAKE3/blake3_sse2_x86-64_unix.S b/llvm/lib/Support/BLAKE3/blake3_sse2_x86-64_unix.S
index d69a1706fefe..5563fc3b5ba0 100644
--- a/llvm/lib/Support/BLAKE3/blake3_sse2_x86-64_unix.S
+++ b/llvm/lib/Support/BLAKE3/blake3_sse2_x86-64_unix.S
@@ -16,19 +16,21 @@
 #define _CET_ENDBR
 #endif
 
-#ifdef __APPLE__
-#define HIDDEN .private_extern
+#if defined(__APPLE__)
+#define HIDDEN(symbol) .private_extern symbol
+#elif defined(__ELF__)
+#define HIDDEN(symbol) .hidden symbol
 #else
-#define HIDDEN .hidden
+#define HIDDEN(symbol)
 #endif
 
 .intel_syntax noprefix
-HIDDEN blake3_hash_many_sse2
-HIDDEN _blake3_hash_many_sse2
-HIDDEN blake3_compress_in_place_sse2
-HIDDEN _blake3_compress_in_place_sse2
-HIDDEN blake3_compress_xof_sse2
-HIDDEN _blake3_compress_xof_sse2
+HIDDEN(blake3_hash_many_sse2)
+HIDDEN(_blake3_hash_many_sse2)
+HIDDEN(blake3_compress_in_place_sse2)
+HIDDEN(_blake3_compress_in_place_sse2)
+HIDDEN(blake3_compress_xof_sse2)
+HIDDEN(_blake3_compress_xof_sse2)
 .global blake3_hash_many_sse2
 .global _blake3_hash_many_sse2
 .global blake3_compress_in_place_sse2
diff --git a/llvm/lib/Support/BLAKE3/blake3_sse41_x86-64_unix.S b/llvm/lib/Support/BLAKE3/blake3_sse41_x86-64_unix.S
index c5b103af61c4..5866bfb7ae46 100644
--- a/llvm/lib/Support/BLAKE3/blake3_sse41_x86-64_unix.S
+++ b/llvm/lib/Support/BLAKE3/blake3_sse41_x86-64_unix.S
@@ -16,19 +16,21 @@
 #define _CET_ENDBR
 #endif
 
-#ifdef __APPLE__
-#define HIDDEN .private_extern
+#if defined(__APPLE__)
+#define HIDDEN(symbol) .private_extern symbol
+#elif defined(__ELF__)
+#define HIDDEN(symbol) .hidden symbol
 #else
-#define HIDDEN .hidden
+#define HIDDEN(symbol)
 #endif
 
 .intel_syntax noprefix
-HIDDEN blake3_hash_many_sse41
-HIDDEN _blake3_hash_many_sse41
-HIDDEN blake3_compress_in_place_sse41
-HIDDEN _blake3_compress_in_place_sse41
-HIDDEN blake3_compress_xof_sse41
-HIDDEN _blake3_compress_xof_sse41
+HIDDEN(blake3_hash_many_sse41)
+HIDDEN(_blake3_hash_many_sse41)
+HIDDEN(blake3_compress_in_place_sse41)
+HIDDEN(_blake3_compress_in_place_sse41)
+HIDDEN(blake3_compress_xof_sse41)
+HIDDEN(_blake3_compress_xof_sse41)
 .global blake3_hash_many_sse41
 .global _blake3_hash_many_sse41
 .global blake3_compress_in_place_sse41
diff --git a/clang/BUILD.bazel b/clang/BUILD.bazel
index dccc72267873..ae6d4d28d3a8 100644
--- a/clang/BUILD.bazel
+++ b/clang/BUILD.bazel
@@ -2,6 +2,8 @@
 # See https://llvm.org/LICENSE.txt for license information.
 # SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
 
+load("@bazel_skylib//rules:run_binary.bzl", "run_binary")
+load("@bazel_skylib//rules:write_file.bzl", "write_file")
 load("@rules_cc//cc:defs.bzl", "cc_binary", "cc_library")
 load("@rules_python//python:defs.bzl", "py_binary")
 load(
@@ -562,23 +564,18 @@ exports_files(
     glob(["include/**/*.td"]),
 )
 
-genrule(
+write_file(
     name = "basic_version_gen",
-    outs = ["include/clang/Basic/Version.inc"],
-    cmd = (
-        "echo '#define CLANG_VERSION {vers}' >> $@\n" +
-        "echo '#define CLANG_VERSION_MAJOR {major}' >> $@\n" +
-        "echo '#define CLANG_VERSION_MAJOR_STRING \"{major}\"' >> $@\n" +
-        "echo '#define CLANG_VERSION_MINOR {minor}' >> $@\n" +
-        "echo '#define CLANG_VERSION_PATCHLEVEL {patch}' >> $@\n" +
-        "echo '#define MAX_CLANG_ABI_COMPAT_VERSION {major}' >> $@\n" +
-        "echo '#define CLANG_VERSION_STRING \"{vers}\"' >> $@\n"
-    ).format(
-        major = LLVM_VERSION_MAJOR,
-        minor = LLVM_VERSION_MINOR,
-        patch = LLVM_VERSION_PATCH,
-        vers = PACKAGE_VERSION,
-    ),
+    out = "include/clang/Basic/Version.inc",
+    content = [
+        "#define CLANG_VERSION {}".format(PACKAGE_VERSION),
+        "#define CLANG_VERSION_MAJOR {}".format(LLVM_VERSION_MAJOR),
+        "#define CLANG_VERSION_MAJOR_STRING \"{}\"".format(LLVM_VERSION_MAJOR),
+        "#define CLANG_VERSION_MINOR {}".format(LLVM_VERSION_MINOR),
+        "#define CLANG_VERSION_PATCHLEVEL {}".format(LLVM_VERSION_PATCH),
+        "#define MAX_CLANG_ABI_COMPAT_VERSION {}".format(LLVM_VERSION_MAJOR),
+        "#define CLANG_VERSION_STRING \"{}\"".format(PACKAGE_VERSION),
+    ],
 )
 
 cc_library(
@@ -596,13 +593,15 @@ cc_library(
 
 # TODO: This should get replaced with something that actually generates the
 # correct version number.
-genrule(
+write_file(
     name = "vcs_version_gen",
     # This should be under lib/Basic, but because of how the include paths
     # are passed through bazel, it's easier to drop generated files next to
     # the other includes.
-    outs = ["include/VCSVersion.inc"],
-    cmd = "echo '#undef CLANG_REVISION' > $@",
+    out = "include/VCSVersion.inc",
+    content = [
+        "#undef CLANG_REVISION",
+    ],
 )
 
 py_binary(
@@ -964,16 +963,21 @@ cc_library(
     ],
 )
 
-genrule(
+run_binary(
     name = "analysis_htmllogger_gen",
     srcs = [
-        "lib/Analysis/FlowSensitive/HTMLLogger.html",
         "lib/Analysis/FlowSensitive/HTMLLogger.css",
+        "lib/Analysis/FlowSensitive/HTMLLogger.html",
         "lib/Analysis/FlowSensitive/HTMLLogger.js",
     ],
     outs = ["lib/Analysis/FlowSensitive/HTMLLogger.inc"],
-    cmd = "$(location :bundle_resources) $@ $(SRCS)",
-    tools = [":bundle_resources"],
+    args = [
+        "$(execpath lib/Analysis/FlowSensitive/HTMLLogger.inc)",
+        "$(execpath lib/Analysis/FlowSensitive/HTMLLogger.html)",
+        "$(execpath lib/Analysis/FlowSensitive/HTMLLogger.css)",
+        "$(execpath lib/Analysis/FlowSensitive/HTMLLogger.js)",
+    ],
+    tool = ":bundle_resources",
 )
 
 cc_library(
@@ -1511,7 +1515,9 @@ cc_library(
         "lib/Driver",
     ],
     linkopts = select({
-        "@platforms//os:windows": ["version.lib"],
+        "//llvm:is_windows_clang_mingw": ["-lversion"],
+        "//llvm:is_windows_clang_cl": ["version.lib"],
+        "//llvm:is_windows_msvc": ["version.lib"],
         "//conditions:default": [],
     }),
     textual_hdrs = glob([
diff --git a/lld/BUILD.bazel b/lld/BUILD.bazel
index 722b695e3bf2..acd80968285b 100644
--- a/lld/BUILD.bazel
+++ b/lld/BUILD.bazel
@@ -2,7 +2,8 @@
 # See https://llvm.org/LICENSE.txt for license information.
 # SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
 
-load("@rules_cc//cc:cc_library.bzl", "cc_library")
+load("@bazel_skylib//rules:write_file.bzl", "write_file")
+load("@rules_cc//cc:defs.bzl", "cc_library")
 load(
     "//:vars.bzl",
     "LLVM_VERSION",
@@ -19,17 +20,21 @@ package(
 licenses(["notice"])
 
 # TODO: Actually compute version info
-genrule(
+write_file(
     name = "config_version_gen",
-    outs = ["include/lld/Common/Version.inc"],
-    cmd = "echo '#define LLD_VERSION_STRING \"{}\"' > $@".format(LLVM_VERSION),
+    out = "include/lld/Common/Version.inc",
+    content = [
+        "#define LLD_VERSION_STRING \"{}\"".format(LLVM_VERSION),
+    ],
 )
 
-genrule(
+write_file(
     name = "vcs_version_gen",
-    outs = ["Common/VCSVersion.inc"],
-    cmd = "echo '#undef LLD_REVISION' >> $@\n" +
-          "echo '#undef LLD_REPOSITORY' >> $@\n",
+    out = "Common/VCSVersion.inc",
+    content = [
+        "#undef LLD_REVISION",
+        "#undef LLD_REPOSITORY",
+    ],
 )
 
 # See https://github.com/bazelbuild/bazel/issues/13803
diff --git a/llvm/BUILD.bazel b/llvm/BUILD.bazel
index a8ec86a7500d..b9ed7e4e8621 100644
--- a/llvm/BUILD.bazel
+++ b/llvm/BUILD.bazel
@@ -6,6 +6,7 @@ load("@rules_cc//cc:cc_binary.bzl", "cc_binary")
 load("@rules_cc//cc:cc_library.bzl", "cc_library")
 load("@bazel_skylib//rules:common_settings.bzl", "string_flag")
 load("@bazel_skylib//rules:expand_template.bzl", "expand_template")
+load("@bazel_skylib//rules:write_file.bzl", "write_file")
 load("@rules_python//python:defs.bzl", "py_binary")
 load("@rules_shell//shell:sh_binary.bzl", "sh_binary")
 load("//mlir:tblgen.bzl", "gentbl_cc_library", "gentbl_filegroup", "td_library")
@@ -209,18 +210,40 @@ cc_library(
     deps = [":config"],
 )
 
-genrule(
+write_file(
     name = "generate_vcs_revision",
-    outs = ["include/llvm/Support/VCSRevision.h"],
-    cmd = "echo '#undef LLVM_REVISION' >> $@\n" +
-          "echo '#undef LLVM_REPOSITORY' >> $@\n",
+    out = "include/llvm/Support/VCSRevision.h",
+    content = [
+        "#undef LLVM_REVISION",
+        "#undef LLVM_REPOSITORY",
+    ],
 )
 
-genrule(
+write_file(
     name = "generate_static_extension_registry",
-    outs = ["include/llvm/Support/Extension.def"],
-    cmd = "echo -e '// extension handlers' >> $@\n" +
-          "echo -e '#undef HANDLE_EXTENSION' >> $@\n",
+    out = "include/llvm/Support/Extension.def",
+    content = [
+        "// extension handlers",
+        "#undef HANDLE_EXTENSION",
+    ],
+)
+
+config_setting(
+    name = "is_windows_clang_mingw",
+    constraint_values = ["@platforms//os:windows"],
+    flag_values = {"@rules_cc//cc/compiler:compiler": "clang"},
+)
+
+config_setting(
+    name = "is_windows_clang_cl",
+    constraint_values = ["@platforms//os:windows"],
+    flag_values = {"@rules_cc//cc/compiler:compiler": "clang-cl"},
+)
+
+config_setting(
+    name = "is_windows_msvc",
+    constraint_values = ["@platforms//os:windows"],
+    flag_values = {"@rules_cc//cc/compiler:compiler": "msvc-cl"},
 )
 
 cc_library(
@@ -297,7 +320,21 @@ cc_library(
     }),
     includes = ["include"],
     linkopts = select({
-        "@platforms//os:windows": [
+        ":is_windows_clang_mingw": [
+            "-lole32",
+            "-luuid",
+            "-lws2_32",
+            "-lntdll",
+        ],
+        ":is_windows_clang_cl": [
+            "ole32.lib",
+            "uuid.lib",
+            "ws2_32.lib",
+            "ntdll.lib",
+        ],
+        ":is_windows_msvc": [
+            "ole32.lib",
+            "uuid.lib",
             "ws2_32.lib",
             "ntdll.lib",
         ],
