diff --git a/lld/COFF/COFFLinkerContext.h b/lld/COFF/COFFLinkerContext.h
index 0f8f2062b9f2..a2273e0ad9ae 100644
--- a/lld/COFF/COFFLinkerContext.h
+++ b/lld/COFF/COFFLinkerContext.h
@@ -60,6 +60,7 @@ public:
   }
 
   std::vector<ObjFile *> objFileInstances;
+  std::vector<BitcodeFile *> lazyBitcodeFileInstances;
   std::map<std::string, PDBInputFile *> pdbInputFileInstances;
   std::vector<ImportFile *> importFileInstances;
   std::int64_t consumedInputsSize = 0;
diff --git a/lld/COFF/Driver.cpp b/lld/COFF/Driver.cpp
index ec1875638030..132138dcf03f 100644
--- a/lld/COFF/Driver.cpp
+++ b/lld/COFF/Driver.cpp
@@ -12,6 +12,7 @@
 #include "DebugTypes.h"
 #include "ICF.h"
 #include "InputFiles.h"
+#include "LTO.h"
 #include "MarkLive.h"
 #include "MinGW.h"
 #include "SymbolTable.h"
@@ -205,10 +206,12 @@ static bool compatibleMachineType(COFFLinkerContext &ctx, MachineTypes mt) {
 void LinkerDriver::addFile(InputFile *file) {
   Log(ctx) << "Reading " << toString(file);
   if (file->lazy) {
-    if (auto *f = dyn_cast<BitcodeFile>(file))
+    if (auto *f = dyn_cast<BitcodeFile>(file)) {
+      ctx.lazyBitcodeFileInstances.push_back(f);
       f->parseLazy();
-    else
+    } else {
       cast<ObjFile>(file)->parseLazy();
+    }
   } else {
     ctx.consumedInputsSize += file->mb.getBufferSize();
     file->parse();
@@ -2757,6 +2760,8 @@ void LinkerDriver::linkerMain(ArrayRef<const char *> argsArr) {
   // Do LTO by compiling bitcode input files to a set of native COFF files then
   // link those files (unless -thinlto-index-only was given, in which case we
   // resolve symbols and write indices, but don't generate native code or link).
+  if (config->thinLTOIndexOnly)
+    thinLTOCreateEmptyIndexFiles(ctx);
   ltoCompilationDone = true;
   ctx.forEachSymtab([](SymbolTable &symtab) { symtab.compileBitcodeFiles(); });
 
diff --git a/lld/COFF/LTO.cpp b/lld/COFF/LTO.cpp
index d55f95493a85..48faed722db7 100644
--- a/lld/COFF/LTO.cpp
+++ b/lld/COFF/LTO.cpp
@@ -150,6 +150,33 @@ BitcodeCompiler::BitcodeCompiler(COFFLinkerContext &c) : ctx(c) {
 
 BitcodeCompiler::~BitcodeCompiler() = default;
 
+// If a lazy BitcodeFile has not been added to the link, emit an empty index
+// file. This matches ELF and GNU gold plugin behavior for distributed builds.
+void lld::coff::thinLTOCreateEmptyIndexFiles(COFFLinkerContext &ctx) {
+  DenseSet<StringRef> linkedBitcodeFiles;
+  ctx.forEachSymtab([&](SymbolTable &symtab) {
+    for (BitcodeFile *f : symtab.bitcodeFileInstances)
+      linkedBitcodeFiles.insert(f->getName());
+  });
+
+  for (BitcodeFile *f : ctx.lazyBitcodeFileInstances) {
+    if (!f->lazy || linkedBitcodeFiles.contains(f->getName()))
+      continue;
+    std::string path = lto::getThinLTOOutputFile(
+        f->obj->getName(), ctx.config.thinLTOPrefixReplaceOld,
+        ctx.config.thinLTOPrefixReplaceNew);
+    std::unique_ptr<raw_fd_ostream> os = openFile(path + ".thinlto.bc");
+    if (!os)
+      continue;
+
+    ModuleSummaryIndex m(/*HaveGVs=*/false);
+    m.setSkipModuleByDistributedBackend();
+    writeIndexToFile(m, *os);
+    if (ctx.config.thinLTOEmitImportsFiles)
+      openFile(path + ".imports");
+  }
+}
+
 static void undefine(Symbol *s) { replaceSymbol<Undefined>(s, s->getName()); }
 
 void BitcodeCompiler::add(BitcodeFile &f) {
diff --git a/lld/COFF/LTO.h b/lld/COFF/LTO.h
index 6826251b5ffa..1fbf0196347d 100644
--- a/lld/COFF/LTO.h
+++ b/lld/COFF/LTO.h
@@ -38,6 +38,8 @@ class BitcodeFile;
 class InputFile;
 class COFFLinkerContext;
 
+void thinLTOCreateEmptyIndexFiles(COFFLinkerContext &ctx);
+
 class BitcodeCompiler {
 public:
   BitcodeCompiler(COFFLinkerContext &ctx);
