Subject: [PATCH] blake3.patch 3
blake3.patch 2
blake3.patch
---
Index: llvm/BUILD.bazel
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/llvm/BUILD.bazel b/llvm/BUILD.bazel
--- a/llvm/BUILD.bazel	(revision ce6b71a665d55db6d6329a254092226eb030cbe8)
+++ b/llvm/BUILD.bazel	(revision acfd66b043cfa923097dff1679190459a4fd580d)
@@ -2,6 +2,7 @@
 # See https://llvm.org/LICENSE.txt for license information.
 # SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception

+load("@bazel_skylib//lib:selects.bzl", "selects")
 load("@bazel_skylib//rules:common_settings.bzl", "string_flag")
 load("@bazel_skylib//rules:expand_template.bzl", "expand_template")
 load("@bazel_skylib//rules:write_file.bzl", "write_file")
@@ -244,6 +245,57 @@
     flag_values = {"@rules_cc//cc/compiler:compiler": "msvc-cl"},
 )

+config_setting(
+    name = "is_x86_64_windows_clang_mingw",
+    constraint_values = [
+        "@platforms//cpu:x86_64",
+        "@platforms//os:windows",
+    ],
+    flag_values = {"@rules_cc//cc/compiler:compiler": "clang"},
+)
+
+# TODO(zbarsky): Is there a better way to express this?
+config_setting(
+    name = "_stamp_binary_on_check",
+    values = {"stamp": "1"},
+)
+
+config_setting(
+    name = "_stamp_binary_off_check",
+    values = {"stamp": "0"},
+)
+
+selects.config_setting_group(
+    name = "_true",
+    match_any = [
+        ":_stamp_binary_on_check",
+        ":_stamp_binary_off_check",
+    ],
+)
+
+alias(
+    name = "non_windows",
+    actual = select({
+        "@platforms//os:windows": "@platforms//os:macos",
+        "//conditions:default": ":_true",
+    }),
+)
+
+selects.config_setting_group(
+    name = "is_x86_64_non_windows",
+    match_all = [
+        "@platforms//cpu:x86_64",
+        ":non_windows",
+    ],
+)
+
+BLAKE3_x86_64_ASM_SOURCE_PATTERNS = [
+    "lib/Support/BLAKE3/blake3_avx2_x86-64_%s.S",
+    "lib/Support/BLAKE3/blake3_avx512_x86-64_%s.S",
+    "lib/Support/BLAKE3/blake3_sse2_x86-64_%s.S",
+    "lib/Support/BLAKE3/blake3_sse41_x86-64_%s.S",
+]
+
 cc_library(
     name = "Support",
     srcs = glob([
@@ -272,11 +324,13 @@
         "@platforms//cpu:aarch64": [
             "lib/Support/BLAKE3/blake3_neon.c",
         ],
-        "@platforms//cpu:x86_64": [
-            "lib/Support/BLAKE3/blake3_avx2_x86-64_unix.S",
-            "lib/Support/BLAKE3/blake3_avx512_x86-64_unix.S",
-            "lib/Support/BLAKE3/blake3_sse2_x86-64_unix.S",
-            "lib/Support/BLAKE3/blake3_sse41_x86-64_unix.S",
+        ":is_x86_64_windows_clang_mingw": [
+            pattern % "windows_gnu"
+            for pattern in BLAKE3_x86_64_ASM_SOURCE_PATTERNS
+        ],
+        ":is_x86_64_non_windows": [
+            pattern % "unix"
+            for pattern in BLAKE3_x86_64_ASM_SOURCE_PATTERNS
         ],
         "//conditions:default": [
         ],
