commit e63c26093fddb963a94b6207beaaf576c2c54115
Author: David Zbarsky <dzbarsky@gmail.com>
Date:   Sat Dec 13 17:05:57 2025 -0500

    Remove builtin headers from frontend target

diff --git a/clang/BUILD.bazel b/clang/BUILD.bazel
index 5ceb2e42c23d..c7ba4fed1760 100644
--- a/clang/BUILD.bazel
+++ b/clang/BUILD.bazel
@@ -1744,7 +1744,7 @@ cc_library(
         "include/clang/Frontend/*.h",
     ]),
     copts = ["$(STACK_FRAME_UNLIMITED)"],
-    data = [":builtin_headers_gen"],
+    #data = [":builtin_headers_gen"],
     includes = ["include"],
     deps = [
         ":apinotes",
